libs/ngx-sherfire/src/lib/firestore/derivable-firestore.service.ts

```diff
@@ -0,0 +1,136 @@
+    }
+
+    /**
+     * Build a derivable based on a Query or CollectionReference that returns a stream of arrays of document contents.
+     */
+    data$<App, Db extends DocumentData>(query: Query<App, Db>): Derivable<App[]>;
+    /**
+     * Build a derivable based on a DocumentReference that returns a stream of document contents. The returned derivable is settable and
+     * setting a value to the derivable will update the firestore document.
+     */
+    data$<App, Db extends DocumentData>(ref: DocumentReference<App, Db>): SettableDerivable<App>;
+    data$<App, Db extends DocumentData>(
+        v: DocumentReference<App, Db> | Query<App, Db>,
+    ): SettableDerivable<App> | Derivable<App[]> {
+        return v instanceof DocumentReference
+            ? this.CACHES.doc.data(v)
+            : v instanceof CollectionReference
+            ? this.CACHES.collection.data(v)
+            : this.queryData$(v);
+    }
+
+    private readonly CACHES = this.buildCaches();
+
+    private querySnapshot$<AppModelType, DbModelType extends DocumentData>(
+        query: Query<AppModelType, DbModelType>,
+    ): Derivable<QuerySnapshot<AppModelType, DbModelType>> {
+        return fromEventPattern<QuerySnapshot<AppModelType, DbModelType>>(value$ =>
+            onSnapshot(query, ...this.snapshotEventHandlers(value$)),
+        );
+    }
+
+    private docSnapshot$<AppModelType, DbModelType extends DocumentData>(
+        ref: DocumentReference<AppModelType, DbModelType>,
+    ): Derivable<DocumentSnapshot<AppModelType, DbModelType>> {
+        return fromEventPattern<DocumentSnapshot<AppModelType, DbModelType>>(value$ =>
+            onSnapshot(ref, ...this.snapshotEventHandlers(value$)),
+        );
+    }
+
+    private snapshotEventHandlers<T>(value$: DerivableAtom<T>) {
+        return [
+            (snap: T) => this.zone.run(() => value$.set(snap)),
+            // istanbul ignore next, because I don't know how to force that to happen
+            (err: FirestoreError) => this.zone.run(() => value$.setFinal(error(err))),
+        ] as const;
+    }
+
+    private queryDocs$<AppModelType, DbModelType extends DocumentData>(
+        query: Query<AppModelType, DbModelType>,
+    ): Derivable<Array<QueryDocumentSnapshot<AppModelType, DbModelType>>> {
+        return this.snapshot$(query).map(s => s.docs);
+    }
+
+    private queryData$<AppModelType, DbModelType extends DocumentData>(
+        query: Query<AppModelType, DbModelType>,
+    ): Derivable<AppModelType[]> {
+        return this.docs$(query).map(snaps => snaps.map(snap => snap.data()));
+    }
+
+    private buildCaches() {
+        const CACHE_OPTIONS: DerivableCacheOptions<
+            DocumentReference<any, any> | CollectionReference<any, any>,
+            string
+        > = {
+            delayedEviction: true,
+            mapKeys: v => v.path,
+        };
+
+        return {
+            collection: {
```

Comment:
the same as what used to happen in `augmentQueries` when used with a `CollectionReference` (and not a `Query` as these cannot be cached)